spark/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala

```diff
@@ -583,6 +588,191 @@ trait DataSkippingReaderBase
+      //    We shouldn't apply partition-like filters in this case, but
+      //    rewriteDataFiltersAsPartitionLikeInternal validates the column is not a Timestamp,
+      //    so we don't have to check here.
+      // 2. The min-max stats on a string column might be truncated for an unpartitioned table.
+      //    Note that just validating that the min and max are equal is enough to prevent this case
+      //    - if the string is truncated, the collected max value is guaranteed to be longer than
+      //    the min value due to the tiebreaker character(s) appended at the end of the max.
+      Not(
+        Or(
+          allNulls,
+          And(
+            EqualNullSafe(
+              resolvedPartitionLikeReference.minExpr, resolvedPartitionLikeReference.maxExpr),
+            EqualTo(resolvedPartitionLikeReference.nullCountExpr, Literal(0L))
+          )
+        )
+      )
+    }
+
+    /**
+     * Rewrites the given expression as a partition-like expression if possible:
+     * 1. Rewrite the attribute references in the expression to reference the collected min stats
+     *     on the attribute reference's column.
+     * 2. Construct an expression that returns true if any of the referenced columns are not
+     *     partition-like on a given file.
+     * The rewritten expression is a union of the above expressions: a file is read if it's either
+     * not partition-like on any of the columns or if the rewritten expression evaluates to true.
+     *
+     * @param clusteringColumns   The columns that are used for clustering.
+     * @param expr                The data filtering expression to rewrite.
+     * @return                    If the expression is safe to rewrite, return the rewritten
+     *                            expression. Otherwise, return None.
+     */
+    def rewriteDataFiltersAsPartitionLike(
+        clusteringColumns: Seq[String], expr: Expression): Option[DataSkippingPredicate] = {
+      val clusteringColumnPaths =
+        clusteringColumns.map(UnresolvedAttribute.quotedString(_).nameParts).toSet
+      rewriteDataFiltersAsPartitionLikeInternal(expr, clusteringColumnPaths).map {
+        case (newExpr, referencedStats) =>
+          // Create an expression that returns true if a file must be read because it has mismatched
+          // min-max values or partial nulls on any of the referenced columns.
+          val numRecordsStatsCol = StatsColumn(NUM_RECORDS, pathToColumn = Nil, LongType)
+          val numRecordsColOpt = getStatsColumnOpt(numRecordsStatsCol)
+          val (finalExpr, statsCols) =
+            referencedStats.foldLeft((newExpr, Seq(numRecordsStatsCol))) {
+              case ((oldExpr, stats), resolvedReference) =>
+                val updatedExpr = Or(
+                  oldExpr, fileMustBeScanned(resolvedReference, numRecordsColOpt))
+                (updatedExpr, stats ++ resolvedReference.referencedStatsCols)
```

Comment:
Seems like we should be doing `++=` on a mutable structure, to avoid quadratic behavior from copy-on-write the list over and over?

Reply:
At the moment, the quadratic behavior shouldn't be too much of a concern (given that the returned references is a set of clustering columns, this set can have a maximum size of 4). However, in case we do expand coverage in the future, this quadratic behavior would be easy to miss. Updated to use a mutable buffer to avoid this case.